File: go/ycsb/workload.go
package ycsb

import (
	"fmt"
	"math/rand/v2"
	"strings"
)

// Workload represents one of the standard YCSB core workloads (A through F).
type Workload string

const (
	// WorkloadA represents Workload A (50% Read, 50% Update).
	WorkloadA Workload = "A"
	// WorkloadB represents Workload B (95% Read, 5% Update).
	WorkloadB Workload = "B"
	// WorkloadC represents Workload C (100% Read).
	WorkloadC Workload = "C"
	// WorkloadD represents Workload D (95% Read Latest, 5% Insert).
	WorkloadD Workload = "D"
	// WorkloadE represents Workload E (95% Short Range Scan, 5% Insert).
	WorkloadE Workload = "E"
	// WorkloadF represents Workload F (50% Read, 50% Read-Modify-Write).
	WorkloadF Workload = "F"
)

// ParseWorkload parses a workload identifier string into a Workload type.
func ParseWorkload(s string) (Workload, error) {
	switch strings.ToUpper(strings.TrimSpace(s)) {
	case "A":
		return WorkloadA, nil
	case "B":
		return WorkloadB, nil
	case "C":
		return WorkloadC, nil
	case "D":
		return WorkloadD, nil
	case "E":
		return WorkloadE, nil
	case "F":
		return WorkloadF, nil
	default:
		return "", fmt.Errorf("unknown YCSB workload %q (expected A, B, C, D, E, or F)", s)
	}
}

// KeyDistribution represents the key selection distribution.
type KeyDistribution string

const (
	// DistributionScrambledZipfian represents the Scrambled Zipfian distribution (default).
	DistributionScrambledZipfian KeyDistribution = "scrambled-zipfian"
	// DistributionZipfian represents the standard Zipfian distribution.
	DistributionZipfian KeyDistribution = "zipfian"
	// DistributionUniform represents the uniform random distribution.
	DistributionUniform KeyDistribution = "uniform"
)

// ParseDistribution parses a distribution string into a KeyDistribution type.
func ParseDistribution(s string) (KeyDistribution, error) {
	switch strings.ToLower(strings.TrimSpace(s)) {
	case "scrambled-zipfian", "scrambledzipfian", "scrambled_zipfian":
		return DistributionScrambledZipfian, nil
	case "zipfian":
		return DistributionZipfian, nil
	case "uniform":
		return DistributionUniform, nil
	default:
		return "", fmt.Errorf("unknown key distribution %q (expected scrambled-zipfian, zipfian, or uniform)", s)
	}
}

// Operation represents an individual YCSB transaction operation type.
type Operation int

const (
	// OperationRead performs a point lookup for a single record.
	OperationRead Operation = iota
	// OperationUpdate performs a blind write updating a random field in a record.
	OperationUpdate
	// OperationInsert inserts a brand new record with all fields populated.
	OperationInsert
	// OperationScan performs a short range scan starting at a key.
	OperationScan
	// OperationReadModifyWrite reads a record and writes back an updated field atomically.
	OperationReadModifyWrite
)

// String returns the string representation of the operation.
func (op Operation) String() string {
	switch op {
	case OperationRead:
		return "READ"
	case OperationUpdate:
		return "UPDATE"
	case OperationInsert:
		return "INSERT"
	case OperationScan:
		return "SCAN"
	case OperationReadModifyWrite:
		return "RMW"
	default:
		return "UNKNOWN"
	}
}

// ChooseOperation selects an operation according to the specified YCSB workload distribution.
func ChooseOperation(workload Workload) Operation {
	switch workload {
	case WorkloadA:
		// 50% Read, 50% Update
		if rand.Float64() < 0.5 {
			return OperationRead
		}
		return OperationUpdate

	case WorkloadB:
		// 95% Read, 5% Update
		if rand.Float64() < 0.95 {
			return OperationRead
		}
		return OperationUpdate

	case WorkloadC:
		// 100% Read
		return OperationRead

	case WorkloadD:
		// 95% Read (latest), 5% Insert
		if rand.Float64() < 0.95 {
			return OperationRead
		}
		return OperationInsert

	case WorkloadE:
		// 95% Scan, 5% Insert
		if rand.Float64() < 0.95 {
			return OperationScan
		}
		return OperationInsert

	case WorkloadF:
		// 50% Read, 50% Read-Modify-Write
		if rand.Float64() < 0.5 {
			return OperationRead
		}
		return OperationReadModifyWrite

	default:
		return OperationRead
	}
}

File: rust/src/ycsb/workload.rs
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use clap::ValueEnum;
use rand::{RngExt, rng};

/// Standard YCSB core workload definitions (Workloads A through F).
#[derive(ValueEnum, Clone, Copy, Debug, PartialEq, Eq)]
pub enum YcsbWorkload {
    /// Workload A: 50% Read, 50% Update (Heavy Update).
    #[value(name = "A", alias = "a")]
    A,
    /// Workload B: 95% Read, 5% Update (Read Mostly).
    #[value(name = "B", alias = "b")]
    B,
    /// Workload C: 100% Read (Read Only).
    #[value(name = "C", alias = "c")]
    C,
    /// Workload D: 95% Read, 5% Insert (Read Latest).
    #[value(name = "D", alias = "d")]
    D,
    /// Workload E: 95% Scan, 5% Insert (Short Scans).
    #[value(name = "E", alias = "e")]
    E,
    /// Workload F: 50% Read, 50% Read-Modify-Write (Transactional Atomic Update).
    #[value(name = "F", alias = "f")]
    F,
}

impl YcsbWorkload {
    /// Returns the uppercase string name of the workload (e.g. `"A"`, `"B"`).
    pub fn name(&self) -> &'static str {
        match self {
            Self::A => "A",
            Self::B => "B",
            Self::C => "C",
            Self::D => "D",
            Self::E => "E",
            Self::F => "F",
        }
    }

    /// Selects the next operation according to the workload's target probability distribution.
    pub(crate) fn next_operation(&self) -> Operation {
        let mut random_number_generator = rng();
        let sample = random_number_generator.random_range(0.0..1.0);

        match self {
            Self::A => {
                // 50% Read, 50% Update
                if sample < 0.50 {
                    Operation::Read
                } else {
                    Operation::Update
                }
            }
            Self::B => {
                // 95% Read, 5% Update
                if sample < 0.95 {
                    Operation::Read
                } else {
                    Operation::Update
                }
            }
            Self::C => {
                // 100% Read
                Operation::Read
            }
            Self::D => {
                // 95% Read, 5% Insert
                if sample < 0.95 {
                    Operation::Read
                } else {
                    Operation::Insert
                }
            }
            Self::E => {
                // 95% Scan, 5% Insert
                if sample < 0.95 {
                    Operation::Scan
                } else {
                    Operation::Insert
                }
            }
            Self::F => {
                // 50% Read, 50% Read-Modify-Write
                if sample < 0.50 {
                    Operation::Read
                } else {
                    Operation::ReadModifyWrite
                }
            }
        }
    }
}

/// Key request distribution strategy used across YCSB operations.
#[derive(ValueEnum, Clone, Copy, Debug, PartialEq, Eq)]
pub enum KeyDistribution {
    /// Standard Zipfian distribution centered at key 0.
    #[value(name = "zipfian", alias = "ZIPFIAN")]
    Zipfian,
    /// Uniformly distributed random keys across the entire key space.
    #[value(name = "uniform", alias = "UNIFORM")]
    Uniform,
    /// Scrambled Zipfian distribution scattering hot keys across the entire key space using FNV-1a.
    #[value(
        name = "scrambled-zipfian",
        alias = "scrambled_zipfian",
        alias = "SCRAMBLED_ZIPFIAN",
        alias = "scrambledzipfian"
    )]
    ScrambledZipfian,
}

/// Primitive database operation executed by a YCSB worker.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub(crate) enum Operation {
    /// Single row lookup by primary key.
    Read,
    /// In-place single row column update.
    Update,
    /// New row insertion with monotonic key sequence.
    Insert,
    /// Range scan starting from random key for a given limit.
    Scan,
    /// Transactional read-modify-write on a row.
    ReadModifyWrite,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn workload_c_always_reads() -> anyhow::Result<()> {
        for _ in 0..1000 {
            assert_eq!(
                YcsbWorkload::C.next_operation(),
                Operation::Read,
                "Workload C must always return Read operation"
            );
        }
        Ok(())
    }

    #[test]
    fn workload_a_distribution() -> anyhow::Result<()> {
        let mut read_count = 0;
        let mut update_count = 0;
        let iterations = 10_000;

        for _ in 0..iterations {
            match YcsbWorkload::A.next_operation() {
                Operation::Read => read_count += 1,
                Operation::Update => update_count += 1,
                other => anyhow::bail!("Unexpected operation in Workload A: {:?}", other),
            }
        }

        let read_ratio = read_count as f64 / iterations as f64;
        assert!(
            (read_ratio - 0.50).abs() < 0.03,
            "Workload A read ratio {:.3} should be close to 0.50",
            read_ratio
        );
        assert_eq!(
            read_count + update_count,
            iterations,
            "Total operations must equal iterations"
        );
        Ok(())
    }

    #[test]
    fn workload_b_distribution() -> anyhow::Result<()> {
        let mut read_count = 0;
        let mut update_count = 0;
        let iterations = 10_000;

        for _ in 0..iterations {
            match YcsbWorkload::B.next_operation() {
                Operation::Read => read_count += 1,
                Operation::Update => update_count += 1,
                other => anyhow::bail!("Unexpected operation in Workload B: {:?}", other),
            }
        }

        let read_ratio = read_count as f64 / iterations as f64;
        assert!(
            (read_ratio - 0.95).abs() < 0.02,
            "Workload B read ratio {:.3} should be close to 0.95",
            read_ratio
        );
        assert_eq!(
            read_count + update_count,
            iterations,
            "Total operations must equal iterations"
        );
        Ok(())
    }

    #[test]
    fn workload_d_distribution() -> anyhow::Result<()> {
        let mut read_count = 0;
        let mut insert_count = 0;
        let iterations = 10_000;

        for _ in 0..iterations {
            match YcsbWorkload::D.next_operation() {
                Operation::Read => read_count += 1,
                Operation::Insert => insert_count += 1,
                other => anyhow::bail!("Unexpected operation in Workload D: {:?}", other),
            }
        }

        let read_ratio = read_count as f64 / iterations as f64;
        assert!(
            (read_ratio - 0.95).abs() < 0.02,
            "Workload D read ratio {:.3} should be close to 0.95",
            read_ratio
        );
        assert_eq!(
            read_count + insert_count,
            iterations,
            "Total operations must equal iterations"
        );
        Ok(())
    }

    #[test]
    fn workload_e_distribution() -> anyhow::Result<()> {
        let mut scan_count = 0;
        let mut insert_count = 0;
        let iterations = 10_000;

        for _ in 0..iterations {
            match YcsbWorkload::E.next_operation() {
                Operation::Scan => scan_count += 1,
                Operation::Insert => insert_count += 1,
                other => anyhow::bail!("Unexpected operation in Workload E: {:?}", other),
            }
        }

        let scan_ratio = scan_count as f64 / iterations as f64;
        assert!(
            (scan_ratio - 0.95).abs() < 0.02,
            "Workload E scan ratio {:.3} should be close to 0.95",
            scan_ratio
        );
        assert_eq!(
            scan_count + insert_count,
            iterations,
            "Total operations must equal iterations"
        );
        Ok(())
    }

    #[test]
    fn workload_f_distribution() -> anyhow::Result<()> {
        let mut read_count = 0;
        let mut rmw_count = 0;
        let iterations = 10_000;

        for _ in 0..iterations {
            match YcsbWorkload::F.next_operation() {
                Operation::Read => read_count += 1,
                Operation::ReadModifyWrite => rmw_count += 1,
                other => anyhow::bail!("Unexpected operation in Workload F: {:?}", other),
            }
        }

        let read_ratio = read_count as f64 / iterations as f64;
        assert!(
            (read_ratio - 0.50).abs() < 0.03,
            "Workload F read ratio {:.3} should be close to 0.50",
            read_ratio
        );
        assert_eq!(
            read_count + rmw_count,
            iterations,
            "Total operations must equal iterations"
        );
        Ok(())
    }
}
